### PR TITLE
Fix text editor gets dirty

### PR DESCRIPTION
**Pull request by LawrenceBorst**
Fixes Lundalogik/crm-feature#4099

**How to test:**

I've done my best to test it. Here's what to look for.

Thinking of edge cases, there's only one, which is when the initial value is empty. Try and check if this has any strange behaviour.

Also, the save button should disappear when the value goes back to what it originally was. So if your text editor says "some sentence", and you edit it to something else, and then back to "some sentence", we should not see the save button.

We believe the cause of the original bug was the fact that the default content type was changed from "html" to "markdown". To test, try and change the content type in the card component in lime admin and see if the bug persists.

Make sure saving continues to work as expected.

## Review:
- [ ] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [ ] Commits have the correct *type* for the changes made
- [ ] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [ ] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 1]
```diff
@@ -218,6 +227,13 @@ export class ProsemirrorAdapter {
     private handleTransaction = (transaction) => {
         const newState = this.view.state.apply(transaction);
         this.view.updateState(newState);
+
+        if (isEqual(this.prosemirrorDoc, transaction.doc) && this.shouldEmit) {
+            this.shouldEmit = false;
```
Comment (LawrenceBorst) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:238]: Oh I seem to have the naming backwards 😅

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 2]
```diff
@@ -127,6 +127,10 @@ export class ProsemirrorAdapter {
         ];
     }
 
+    public disconnectedCallback() {
+        this.view.destroy();
```
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:127]: Whoa! Yeah, good catch!

I guess this means we have a memory leak in all apps that have the activities feed enabled with the text-editor… 😬

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 3]
```diff
@@ -122,12 +122,11 @@ export class ProsemirrorAdapter {
...
-        /* eslint-disable multiline-ternary */
-        this.contentConverter =
-            this.contentType === 'markdown'
-                ? new markdownConverter()
-                : new HTMLConverter();
-        /* eslint-enable multiline-ternary */
+        if (this.contentType === 'markdown') {
+            this.contentConverter = new markdownConverter();
+        } else if (this.contentType === 'html') {
+            this.contentConverter = new HTMLConverter();
```
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:134]: Nitpick: This isn't purely a refactor or code style change, since the existing behaviour would default to `HTMLConverter` if `this.contentType` was anything other than `'markdown'`, while this new behaviour is defaulting to `undefined` if `this.contentType` is neither `'markdown'` nor `'html'`.

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 4]
```diff
@@ -122,12 +122,11 @@ export class ProsemirrorAdapter {
     }
 
     private setupContentConverter() {
-        /* eslint-disable multiline-ternary */
-        this.contentConverter =
-            this.contentType === 'markdown'
-                ? new markdownConverter()
-                : new HTMLConverter();
-        /* eslint-enable multiline-ternary */
```
[the remark this answers is not in the document]
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:130]: 😆 
<iframe src="https://giphy.com/embed/jUwpNzg9IcyrK" width="480" height="360" style="" frameBorder="0" class="giphy-embed" allowFullScreen></iframe><p><a href="https://giphy.com/gifs/the-simpsons-scared-homer-simpson-jUwpNzg9IcyrK">via GIPHY</a></p>

[the remark this answers is not in the document]
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:130]: 😆

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 1 above)
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:238]: I think the naming is a little bit confusing here. We seem to only be avoiding emitting something if `this.shouldEmit` is _true_, and then we set it to false, so it will never be true again…?

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:130]: # 🤗

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 5]
```diff
@@ -216,9 +226,20 @@ export class ProsemirrorAdapter {
...
-    private handleTransaction = (transaction) => {
+    private handleTransaction = (transaction: Transaction) => {
         const newState = this.view.state.apply(transaction);
         this.view.updateState(newState);
+
+        if (transaction.getMeta('pointer')) {
+            return;
+        }
+
+        if (isEqual(this.prosemirrorDoc, transaction.doc) && this.shouldEmit) {
```
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:237]: Logical operators are evaluated left-to-right. Since `this.shouldEmit` will _almost always_ be `false`, we can save a lot of work by not recursively comparing the two data structures of unknown size just to then throw the result away when `this.shouldEmit` is evaluated to false.

```suggestion
        if (this.shouldEmit && isEqual(this.prosemirrorDoc, transaction.doc)) {
```

By putting the simple boolean check first, we will never enter the deep equality check if the boolean is `false`.

**Comment by adrianschmidt** [comment]
FYI: I squashed the existing fixup, and rebased on what's currently the latest `main`.

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 6]
```diff
@@ -123,6 +114,15 @@ export class ProsemirrorAdapter {
...
     }
 
+    public componentDidLoad() {
+        // Stencil complains loudly about triggering rerenders in
+        // componentDidLoad, but we have to, so we're using setTimeout to
+        // suppress the warning. /Ads
+        setTimeout(() => {
+            this.initializeTextEditor();
+        }, 0);
+    }
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx]: (I removed this commit.)

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:1]: Oh, btw, the prosemirror-adapter is meant to be an internal component, I assume? Or is it something the consumer would be aware of?

I'm asking, because if it's an internal component that the consumer doesn't know about, we shouldn't use `prosemirror-adapter` as the "scope" for commits that are of a type that show up in the changelog, like `fix`, `feat`, and `perf`. Those should have a scope that is meaningful to the consumer, so the consumer knows which component is affected. So `text-editor` in this case (again, unless I'm mistaken about prosemirror-adapter).

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (LawrenceBorst) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx]: Ah you are right, the style guide always put `render` last :)

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx
Comment (LawrenceBorst) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:1]: This is good to know. I didn't think of this. I'll rename the commits

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 1 above)
Comment (LawrenceBorst) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:241]: Fixed

Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx]: I'm not completely at peace with the fixup you made for this, because we now instead have a flag that says `this.shouldEmit = false` and we're still emitting stuff unless a specific circumstance happens. So I think I'll remove that fixup while I'm fiddling around with some stuff, and then try to find a different name for the property or something.

I'm going to fiddle around with the commits a little bit, so there may be merge conflicts if you are making changes at the same time.

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 7]
```diff
@@ -123,12 +123,11 @@ export class ProsemirrorAdapter {
...
-        this.contentConverter =
-            this.contentType === 'markdown'
-                ? new markdownConverter()
-                : new HTMLConverter();
-        /* eslint-enable multiline-ternary */
+        if (this.contentType === 'markdown') {
+            this.contentConverter = new markdownConverter();
+        } else {
+            this.contentConverter = new HTMLConverter();
+        }
```
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx]: I updated this commit so that there is no change in functionality, and instead changed the `else` into an `else if` in the commit where you added the error logging 👍

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 1 above)
Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:241]: I'm not sure I completely understand how this if statement works 😅

This part specifically:

```ts
isEqual(this.prosemirrorDoc, transaction.doc)
```

As far as I understand, we want to avoid emitting a `change` event if the changes are only syntactic changes that prosemirror makes automatically, stuff like replacing `<b>` with `<strong>`, for example. But I don't understand what `this.prosemirrorDoc` and `transaction.doc` are, or how their values being equal to each other tells us what we need to know? 🤔

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 2 above)
Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx]: This was merged in #3009

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 8]
```diff
@@ -224,11 +226,21 @@ export class ProsemirrorAdapter {
         const tr = this.view.state.tr;
         tr.replaceWith(0, tr.doc.content.size, prosemirrorDoc.content);
         this.view.dispatch(tr);
+        this.suppressChangeEvent = false;
```
Comment (civing) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:229]: Should this really be set here? It seem like a better fit to clear this boolean on row 237. I can't see any relevant events here that would make `handleTransaction` to be called before flipping this boolean

Comment (civing) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:229]: But since this seems to have solved it for you guys I'm probably missing somehting 😅

Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:229]: I've looked at it, and I _think_ I disagree. It feels to me like `handleTransaction` shouldn't just assume that because we suppressed one change event, we should let the next one trough. I think it's better that the function that sets the value also clears it.

With the current state of the code, it shouldn't matter in practice, but I think it feels cleaner for potential future changes to not have `handleTransaction` making assumptions like that.

**Review by john-traas — APPROVED** [review]

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 8 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:229]: I agree with this. For some reason I assumed that the dispatch somehow was hiding some async code but it must not then.
